flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java

```diff
@@ -0,0 +1,342 @@
+                            Tuple2.of(
+                                    MapStateDescriptor.class,
+                                    (StateFactory) ChangelogMapState::create))
+                    .collect(Collectors.toMap(t -> t.f0, t -> t.f1));
+
+    /** delegated keyedStateBackend. */
+    private final AbstractKeyedStateBackend<K> keyedStateBackend;
+
+    /**
+     * This is the cache maintained by the DelegateKeyedStateBackend itself. It is not the same as
+     * the underlying delegated keyedStateBackend. InternalKvState is a delegated state.
+     */
+    private final HashMap<String, InternalKvState<K, ?, ?>> keyValueStatesByName;
+
+    private final ExecutionConfig executionConfig;
+
+    private final TtlTimeProvider ttlTimeProvider;
+
+    /** last accessed partitioned state. */
+    @SuppressWarnings("rawtypes")
+    private InternalKvState lastState;
+
+    /** For caching the last accessed partitioned state. */
+    private String lastName;
+
+    public ChangelogKeyedStateBackend(
+            AbstractKeyedStateBackend<K> keyedStateBackend,
+            ExecutionConfig executionConfig,
+            TtlTimeProvider ttlTimeProvider) {
+        this.keyedStateBackend = keyedStateBackend;
+        this.executionConfig = executionConfig;
+        this.ttlTimeProvider = ttlTimeProvider;
+        this.keyValueStatesByName = new HashMap<>();
+    }
+
+    // -------------------- CheckpointableKeyedStateBackend --------------------------------
+    @Override
+    public KeyGroupRange getKeyGroupRange() {
+        return keyedStateBackend.getKeyGroupRange();
+    }
+
+    @Override
+    public void close() throws IOException {
+        keyedStateBackend.close();
+    }
+
+    @Override
+    public void setCurrentKey(K newKey) {
+        keyedStateBackend.setCurrentKey(newKey);
+    }
+
+    @Override
+    public K getCurrentKey() {
+        return keyedStateBackend.getCurrentKey();
+    }
+
+    @Override
+    public TypeSerializer<K> getKeySerializer() {
+        return keyedStateBackend.getKeySerializer();
+    }
+
+    @Override
+    public <N> Stream<K> getKeys(String state, N namespace) {
+        return keyedStateBackend.getKeys(state, namespace);
+    }
+
+    @Override
+    public <N> Stream<Tuple2<K, N>> getKeysAndNamespaces(String state) {
+        return keyedStateBackend.getKeysAndNamespaces(state);
+    }
+
+    @Override
+    public void dispose() {
+        keyedStateBackend.dispose();
+        lastName = null;
+        lastState = null;
+        keyValueStatesByName.clear();
+    }
+
+    @Override
+    public void registerKeySelectionListener(KeySelectionListener<K> listener) {
+        keyedStateBackend.registerKeySelectionListener(listener);
+    }
+
+    @Override
+    public boolean deregisterKeySelectionListener(KeySelectionListener<K> listener) {
+        return keyedStateBackend.deregisterKeySelectionListener(listener);
+    }
+
+    @Override
+    public <N, S extends State, T> void applyToAllKeys(
```

Comment:
I think this method can be refactored to avoid code duplication by:
1. Remove `AbstractKeyedStateBackend.supportConcurrentModification()`
2. Add an abstract method to iterate over keys to `AbstractKeyedStateBackend` (and override in heap backend)
3. Add `AbstractKeyedStateBackend.applyToAllKeys(Function<..., S> stateFactory, ...)` method
4. By default, use `AbstractKeyedStateBackend.getPartitionedState()`
5. In this class override it and by pass `this::getPartitionedState`
So here it would look like:
```
public void applyToAllKeys(...) {
    keyedStateBackend.applyToAllKeys(namespace, namespaceSerializer, stateDescriptor, function,
           this::getPartitionedState);
```